lisa/tests/hotplug.py

```diff
@@ -269,4 +271,174 @@ def cpuhp_seq(cls, nr_operations, hotpluggable_cpus, max_cpus_off, random_gen):
         for cpu in cur_off_cpus:
             yield cpu, 1
 
-# vim :set tabstop=4 shiftwidth=4 textwidth=80 expandtab
+
+class HotplugRollback(FtraceTestBundle):
+
+    @classmethod
+    def _online(cls, target, cpu, online, verify=True):
+        try:
+            if online:
+                target.hotplug.online(cpu)
+            else:
+                target.hotplug.offline(cpu)
+        except Exception as e:
+            if verify:
+                raise e
+
+    @classmethod
+    def _state_can_fail(cls, target, cpu, state, up):
+        """
+        There are no way of probing the kernel for a list of hotplug states
+        that can fail and for which we can test the rollback. We need therefore
+        to try:
+          - If we can set the state in the kernel 'fail' interface.
+          - If the hotplug is reset actually failing (some states can fail only
+            when going up or down)
+        """
+        try:
+            target.hotplug.fail(cpu, state)
+        except TargetStableError:
+            return False
+
+        try:
+            cls._online(target, cpu, up)
+            target.hotplug.fail(cpu, -1)
+            cls._online(target, cpu, not up)
+            return False
+        except TargetStableError:
+            return True
+
+    @classmethod
+    def _prepare_hotplug(cls, target, cpu, up):
+        target.hotplug.fail(cpu, -1)
+        cls._online(target, cpu, not up)
+
+    @classmethod
+    def _get_states(cls, target, cpu, up):
+        states = target.hotplug.get_states()
+        cls._prepare_hotplug(target, cpu, not up)
+        return [
+                state
+                for state in states if cls._state_can_fail(target, cpu,
+                                                           state, up)
+        ]
+
+    @classmethod
+    def _mark_trace(cls, target, ftrace_coll, start=0,
+                    expected=0, up=False, failing_state=0):
+        target.write_value(
+                ftrace_coll.marker_file,
+                "HotplugRollback: test={} expected={} up={} failing_state={}".format(
+                    int(start), int(expected), up, failing_state),
+                verify=False
+        )
+
+    @classmethod
+    def _test_rollback(cls, target, ftrace_coll, cpu, failing_state, up):
+        cls._prepare_hotplug(target, cpu, up=up)
+        target.hotplug.fail(cpu, failing_state)
+        cls._mark_trace(target, ftrace_coll, start=1, up=up,
+            failing_state=failing_state)
+        cls._online(target, cpu, online=up, verify=False)
+        cls._mark_trace(target, ftrace_coll, start=0)
+
+    @classmethod
+    def _from_target(cls, target, *,
+                     res_dir: ArtifactPath = None) -> 'HotplugRollback':
+        target.hotplug.online_all()
```

Comment:
Given the manipulation of hotplug we do at multiple levels of helpers in that function, it should be encapsulated into something like that:
```
@classmethod
def _from_target(cls, target, *,
                    res_dir: ArtifactPath = None) -> 'HotplugRollback':
    try:
        return cls._do_from_target(target=target, res_dir=res_dir)
    finally:
        target.hotplug.online_all()
        # plus other things potentially to disable hotplug programmatic fail
```